server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpgradeService.java

```diff
@@ -76,90 +70,128 @@ public class MetadataUpgradeService {
+        for (var cursor : metadata.templates()) {
+            String templateName = cursor.key;
+            newMetadata.removeTemplate(templateName);
+            if (templateName == IndexTemplateUpgrader.CRATE_DEFAULTS) {
+                continue;
+            }
+            IndexTemplateMetadata template = cursor.value;
+            RelationName relationName = IndexName.decode(template.name()).toRelationName();
+            RelationMetadata relation = metadata.getRelation(relationName);
+            assert relation == null
+                : "If there is still a template present there shouldn't be any RelationMetadata";
+
+            DocTableInfo docTable = tableFactory.create(template, metadata);
+            Version versionCreated = getFixedVersionCreated(metadata, docTable);
+
+            // versionCreated could be missing from the template settings and "calculated" afterwards
+            // in DocTableInfo, so put it back to RelationMetadata#parameters
+            Settings settings = Settings.builder()
+                // If RelationMetadata exist in the cluster state, make sure to override them with
+                // the upgraded settings which currently takes place on IndexTemplateMetadata
+                .put(pruneArchived(template.settings()))
+                .put(SETTING_VERSION_CREATED, versionCreated)
+                .put(SETTING_VERSION_UPGRADED, Version.CURRENT)
+                .build();
+
+            newMetadata.setTable(
+                versionCreated.before(DocTableInfo.COLUMN_OID_VERSION)
+                    ? NO_OID_COLUMN_OID_SUPPLIER
+                    : newMetadata.columnOidSupplier(),
+                relationName,
+                docTable.allReferences(),
+                settings,
+                docTable.clusteredBy(),
+                docTable.columnPolicy(),
+                docTable.pkConstraintName(),
+                docTable.checkConstraints()
+                    .stream().collect(Collectors.toMap(CheckConstraint::name, CheckConstraint::expressionStr)),
+                docTable.primaryKey(),
+                docTable.partitionedBy(),
+                docTable.isClosed() ? IndexMetadata.State.CLOSE : IndexMetadata.State.OPEN,
+                List.of(),
+                docTable.tableVersion()
+            );
         }
 
         // upgrade index meta data
         for (IndexMetadata indexMetadata : metadata) {
             String indexName = indexMetadata.getIndex().getName();
-            IndexMetadata newMetadata = upgradeIndexMetadata(
+            IndexMetadata newIndexMetadata = upgradeIndexMetadata(
                 indexMetadata,
-                IndexName.isPartitioned(indexName) ?
-                    upgradedIndexTemplateMetadata.get(PartitionName.templateName(indexName)) :
-                    null,
+                IndexName.isPartitioned(indexName)
+                    ? newMetadata.getTemplate(PartitionName.templateName(indexName))
```

Comment:
I just realized that `newMetadata.getTemplate` will now always be `null` with this change.
Which either means that something ought to be broken, or the `upgradeIndexMetadata` can be changed to remove this param - will need to follow up on this